pkg/kubelet/kuberuntime/kuberuntime_sandbox.go

```diff
@@ -0,0 +1,166 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kuberuntime
+
+import (
+	"k8s.io/kubernetes/pkg/api"
+	runtimeApi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+)
+
+// generatePodSandboxConfig generates pod sandbox config from api.Pod.
+func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *api.Pod, podIP string) (*runtimeApi.PodSandboxConfig, error) {
+	_, podName, _ := buildPodName(pod.Name, pod.Namespace, string(pod.UID))
+	podSandboxConfig := &runtimeApi.PodSandboxConfig{
+		Name:        &podName,
+		Labels:      newPodLabels(pod),
+		Annotations: newPodAnnotations(pod),
+	}
+
+	if !kubecontainer.IsHostNetworkPod(pod) {
+		dnsServers, dnsSearches, err := m.runtimeHelper.GetClusterDNS(pod)
+		if err != nil {
+			return nil, err
+		}
+		podSandboxConfig.DnsOptions = &runtimeApi.DNSOption{
+			Servers:  dnsServers,
+			Searches: dnsSearches,
+		}
+
+		// TODO: Add domain support in new runtime interface
+		hostname, _, err := m.runtimeHelper.GeneratePodHostNameAndDomain(pod)
+		if err != nil {
+			return nil, err
+		}
+		podSandboxConfig.Hostname = &hostname
+	}
+
+	cgroupParent := ""
+	podResourceLimits := make(map[api.ResourceName]int64)
+	podResourceRequests := make(map[api.ResourceName]int64)
+	portMappings := []*runtimeApi.PortMapping{}
+	for _, c := range pod.Spec.Containers {
+		for name, limit := range c.Resources.Limits {
+			if l, ok := podResourceLimits[name]; ok {
+				podResourceLimits[name] = l + limit.MilliValue()
+			} else {
+				podResourceLimits[name] = limit.MilliValue()
+			}
+		}
+		for name, req := range c.Resources.Requests {
+			if l, ok := podResourceRequests[name]; ok {
+				podResourceRequests[name] = l + req.MilliValue()
+			} else {
+				podResourceRequests[name] = req.MilliValue()
+			}
+		}
+
+		opts, err := m.runtimeHelper.GenerateRunContainerOptions(pod, &c, podIP)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, port := range opts.PortMappings {
+			hostPort := int32(port.HostPort)
+			containerPort := int32(port.ContainerPort)
+			protocol := toRuntimeProtocol(port.Protocol)
+			portMappings = append(portMappings, &runtimeApi.PortMapping{
+				HostIp:        &port.HostIP,
+				HostPort:      &hostPort,
+				ContainerPort: &containerPort,
+				Protocol:      &protocol,
+				Name:          &port.Name,
+			})
+		}
+
+		cgroupParent = opts.CgroupParent
+	}
+
+	podSandboxConfig.Linux = generatePodSandboxLinuxConfig(pod, cgroupParent)
+
+	if len(portMappings) > 0 {
+		podSandboxConfig.PortMappings = portMappings
+	}
+
+	if len(podResourceLimits) > 0 || len(podResourceRequests) > 0 {
```

Comment:
Let's factor the resources logic to another function, so that we can easily identify them if we go ahead with https://github.com/kubernetes/kubernetes/issues/29871